File: flutter_theoplayer_sdk_android/android/src/main/kotlin/com/theoplayer/theoplayer/TextTrackBridge.kt
package com.theoplayer.theoplayer

import com.theoplayer.android.api.event.EventListener
import com.theoplayer.android.api.event.track.texttrack.AddCueEvent
import com.theoplayer.android.api.event.track.texttrack.ChangeEvent
import com.theoplayer.android.api.event.track.texttrack.CueChangeEvent
import com.theoplayer.android.api.event.track.texttrack.EnterCueEvent
import com.theoplayer.android.api.event.track.texttrack.ExitCueEvent
import com.theoplayer.android.api.event.track.texttrack.RemoveCueEvent
import com.theoplayer.android.api.event.track.texttrack.TextTrackEventTypes
import com.theoplayer.android.api.event.track.texttrack.list.AddTrackEvent
import com.theoplayer.android.api.event.track.texttrack.list.RemoveTrackEvent
import com.theoplayer.android.api.event.track.texttrack.list.TextTrackListEventTypes
import com.theoplayer.android.api.event.track.texttrack.list.TrackListChangeEvent
import com.theoplayer.android.api.event.track.texttrack.texttrackcue.EnterEvent
import com.theoplayer.android.api.event.track.texttrack.texttrackcue.ExitEvent
import com.theoplayer.android.api.event.track.texttrack.texttrackcue.TextTrackCueEventTypes
import com.theoplayer.android.api.event.track.texttrack.texttrackcue.UpdateEvent
import com.theoplayer.android.api.player.Player
import com.theoplayer.android.api.player.track.texttrack.TextTrack
import com.theoplayer.android.api.player.track.texttrack.cue.TextTrackCue
import com.theoplayer.theoplayer.pigeon.THEOplayerFlutterTextTracksAPI
import com.theoplayer.theoplayer.pigeon.THEOplayerNativeTextTracksAPI
import com.theoplayer.theoplayer.pigeon.THEOplayerNativeTextTracksAPI.Companion.setUp
import com.theoplayer.theoplayer.transformers.FlutterTextTrackMode
import com.theoplayer.theoplayer.transformers.TrackTransformer

class TextTrackBridge(
    private val player: Player,
    pigeonMessenger: PigeonBinaryMessengerWrapper,
) : THEOplayerNativeTextTracksAPI {

    private val flutterTextTracksAPI = THEOplayerFlutterTextTracksAPI(pigeonMessenger)

    init {
        setUp(pigeonMessenger, this)
    }

    private val emptyCallback: (Result<Unit>) -> Unit = {}

    private val addTextTrackListener: EventListener<AddTrackEvent> = EventListener {
        flutterTextTracksAPI.onAddTextTrack(
            it.track.id,
            it.track.uid.toLong(),
            it.track.label,
            it.track.language,
            it.track.kind,
            it.track.inBandMetadataTrackDispatchType,
            TrackTransformer.toFlutterTextTrackReadyState(it.track.readyState),
            TrackTransformer.toFlutterTextTrackType(it.track.type),
            it.track.source,
            it.track.isForced,
            TrackTransformer.toFlutterTextTrackMode(it.track.mode),
            emptyCallback
        )
        attachTrackListeners(it.track)
    }

    private val removeTextTrackListener = EventListener<RemoveTrackEvent> {
        removeTrackListeners(it.track)
        flutterTextTracksAPI.onRemoveTextTrack(it.track.uid.toLong(), emptyCallback)
    }

    private val textTrackListChangeListener = EventListener<TrackListChangeEvent> {
        flutterTextTracksAPI.onTextTrackListChange(it.track.uid.toLong(), emptyCallback)
    }

    private val textTrackAddCueListener = EventListener<AddCueEvent> {
        flutterTextTracksAPI.onTextTrackAddCue(it.track.uid.toLong(), it.cue.id, it.cue.uid, it.cue.startTime, it.cue.endTime, it.cue.content.toString(), emptyCallback)
        attachCueListeners(it.track, it.cue)
    }

    private val textTrackRemoveCueListener = EventListener<RemoveCueEvent> {
        removeCueListeners(it.track, it.cue)
        flutterTextTracksAPI.onTextTrackRemoveCue(it.track.uid.toLong(), it.cue.uid, emptyCallback)
    }

    private fun textTrackEnterCueListener(track: TextTrack) = EventListener<EnterCueEvent> {
        flutterTextTracksAPI.onTextTrackEnterCue(track.uid.toLong(), it.cue.uid, emptyCallback)
    }

    private fun textTrackExitCueListener(track: TextTrack) = EventListener<ExitCueEvent> {
        flutterTextTracksAPI.onTextTrackExitCue(track.uid.toLong(), it.cue.uid, emptyCallback)
    }

    private val textTrackCueChangeListener = EventListener<CueChangeEvent> {
        flutterTextTracksAPI.onTextTrackCueChange(it.textTrack.uid.toLong(), emptyCallback)
    }

    private val textTrackChangeListener = EventListener<ChangeEvent> {
        flutterTextTracksAPI.onTextTrackChange(it.track.uid.toLong(), emptyCallback)
    }

    private fun cueEnterListener(track: TextTrack) = EventListener<EnterEvent> {
        flutterTextTracksAPI.onCueEnter(track.uid.toLong(), it.cue.uid, emptyCallback)
    }

    private fun cueExitListener(track: TextTrack) = EventListener<ExitEvent> {
        flutterTextTracksAPI.onCueExit(track.uid.toLong(), it.cue.uid, emptyCallback)
    }

    private fun cueUpdateListener(track: TextTrack) = EventListener<UpdateEvent> {
        flutterTextTracksAPI.onCueUpdate(track.uid.toLong(), it.cue.uid, it.cue.endTime, it.cue.content.toString(), emptyCallback)
    }

    fun attachListeners() {
        player.textTracks.addEventListener(TextTrackListEventTypes.ADDTRACK, addTextTrackListener)
        player.textTracks.addEventListener(TextTrackListEventTypes.REMOVETRACK, removeTextTrackListener)
        player.textTracks.addEventListener(TextTrackListEventTypes.TRACKLISTCHANGE, textTrackListChangeListener)
    }

    fun removeListeners() {
        player.textTracks.removeEventListener(TextTrackListEventTypes.ADDTRACK, addTextTrackListener)
        player.textTracks.removeEventListener(TextTrackListEventTypes.REMOVETRACK, removeTextTrackListener)
        player.textTracks.removeEventListener(TextTrackListEventTypes.TRACKLISTCHANGE, textTrackListChangeListener)

        player.textTracks.forEach { track ->
            removeTrackListeners(track)
        }
    }

    private fun attachTrackListeners(track: TextTrack) {
        track.addEventListener(TextTrackEventTypes.ADDCUE, textTrackAddCueListener)
        track.addEventListener(TextTrackEventTypes.REMOVECUE, textTrackRemoveCueListener)
        track.addEventListener(TextTrackEventTypes.ENTERCUE, textTrackEnterCueListener(track))
        track.addEventListener(TextTrackEventTypes.EXITCUE, textTrackExitCueListener(track))
        track.addEventListener(TextTrackEventTypes.CUECHANGE, textTrackCueChangeListener)
        track.addEventListener(TextTrackEventTypes.CHANGE, textTrackChangeListener)
    }

    private fun removeTrackListeners(track: TextTrack) {
        track.removeEventListener(TextTrackEventTypes.ADDCUE, textTrackAddCueListener)
        track.removeEventListener(TextTrackEventTypes.REMOVECUE, textTrackRemoveCueListener)
        track.removeEventListener(TextTrackEventTypes.ENTERCUE, textTrackEnterCueListener(track))
        track.removeEventListener(TextTrackEventTypes.EXITCUE, textTrackExitCueListener(track))
        track.removeEventListener(TextTrackEventTypes.CUECHANGE, textTrackCueChangeListener)
        track.removeEventListener(TextTrackEventTypes.CHANGE, textTrackChangeListener)
        track.cues?.forEach { cue ->
            removeCueListeners(track, cue)
        }
    }

    private fun attachCueListeners(track: TextTrack, cue: TextTrackCue) {
        cue.addEventListener(TextTrackCueEventTypes.ENTER, cueEnterListener(track))
        cue.addEventListener(TextTrackCueEventTypes.EXIT, cueExitListener(track))
        cue.addEventListener(TextTrackCueEventTypes.UPDATE, cueUpdateListener(track))
    }

    private fun removeCueListeners(track: TextTrack, cue: TextTrackCue) {
        cue.removeEventListener(TextTrackCueEventTypes.ENTER, cueEnterListener(track))
        cue.removeEventListener(TextTrackCueEventTypes.EXIT, cueExitListener(track))
        cue.removeEventListener(TextTrackCueEventTypes.UPDATE, cueUpdateListener(track))
    }

    override fun setMode(textTrackUid: Long, mode: FlutterTextTrackMode) {
        player.textTracks
            .first { it.uid == textTrackUid.toInt() }
            .mode = TrackTransformer.toTextTrackMode(mode)
    }

}

File: flutter_theoplayer_sdk_ios/ios/Classes/TextTrackBridge.swift
//
//  TextTrackBridge.swift
//  flutter_theoplayer_sdk_ios
//
//  Created by Hovig on 15/11/2023.
//

import Foundation
import THEOplayerSDK
import Flutter

class TextTrackBridge: THEOplayerNativeTextTracksAPI {
    
    private let theoplayer: THEOplayer
    private let pigeonMessenger: PigeonBinaryMessengerWrapper
    private let flutterTextTracksAPI: THEOplayerFlutterTextTracksAPI
    
    private var addTextTrackListener: EventListener?
    private var removeTextTrackListener: EventListener?
    private var textTrackListChange: EventListener?
    
    private var textTrackAddCueListener: EventListener?
    private var textTrackRemoveCueListener: EventListener?
    private var textTrackEnterCueListener: EventListener?
    private var textTrackExitCueListener: EventListener?
    private var textTrackCueChangeListener: EventListener?
    private var textTrackChangeListener: EventListener?
    
    private var cueEnterListener: EventListener?
    private var cueExitListener: EventListener?
    private var cueUpdateListener: EventListener?
    
    private let emptyCompletion: (Result<Void, FlutterError>) -> Void = {result in }
    
    init(theoplayer: THEOplayer, pigeonMessenger: PigeonBinaryMessengerWrapper) {
        self.theoplayer = theoplayer
        self.pigeonMessenger = pigeonMessenger
        self.flutterTextTracksAPI = THEOplayerFlutterTextTracksAPI(binaryMessenger: pigeonMessenger)
        THEOplayerNativeTextTracksAPISetup.setUp(binaryMessenger: pigeonMessenger, api: self)
    }
    
    func attachListeners() {
        addTextTrackListener = theoplayer.textTracks.addEventListener(type: TextTrackListEventTypes.ADD_TRACK, listener: { event in
            if let textTrack = event.track as? TextTrack {
                self.flutterTextTracksAPI.onAddTextTrack(
                    id: textTrack.id,
                    uid: Int64(textTrack.uid),
                    label: textTrack.label,
                    language: textTrack.language,
                    kind: textTrack.kind,
                    inBandMetadataTrackDispatchType: textTrack.inBandMetadataTrackDispatchType,
                    readyState: TextTrackReadyState.loaded,
                    type: TextTrackType.none,
                    source: textTrack.src,
                    isForced: false,
                    mode: TrackTransformer.toFlutterTextTrackMode(mode: textTrack.mode),
                    completion: self.emptyCompletion)
                self.attachTrackListeners(track: textTrack)
            }
        })
        
        removeTextTrackListener = theoplayer.textTracks.addEventListener(type: TextTrackListEventTypes.REMOVE_TRACK, listener: { event in
            self.flutterTextTracksAPI.onRemoveTextTrack(uid: Int64(event.track.uid), completion: self.emptyCompletion)
        })
        
        textTrackListChange = theoplayer.textTracks.addEventListener(type: TextTrackListEventTypes.CHANGE, listener: { event in
            self.flutterTextTracksAPI.onTextTrackListChange(uid: Int64(event.track.uid), completion: self.emptyCompletion)
        })
    }
    
    func removeListeners() {
        // TODO: remove force unwraps
        theoplayer.textTracks.removeEventListener(type: TextTrackListEventTypes.ADD_TRACK, listener: addTextTrackListener!)
        theoplayer.textTracks.removeEventListener(type: TextTrackListEventTypes.REMOVE_TRACK, listener: removeTextTrackListener!)
        theoplayer.textTracks.removeEventListener(type: TextTrackListEventTypes.CHANGE, listener: textTrackListChange!)
        
        for i in 0..<theoplayer.textTracks.count {
            let track = theoplayer.textTracks.get(i)
            removeTrackListeners(track: track)
            track.cues.forEach { cue in
                removeCueListeners(track: track, cue: cue)
            }
        }
    }
    
    private func attachTrackListeners(track: TextTrack) {
        textTrackAddCueListener = track.addEventListener(type: TextTrackEventTypes.ADD_CUE, listener: { event in
            self.flutterTextTracksAPI.onTextTrackAddCue(
                textTrackUid: Int64(track.uid),
                id: event.cue.id,
                uid: Int64(event.cue.uid),
                startTime: event.cue.startTime ?? 0,
                endTime: event.cue.endTime ?? 0,
                content: event.cue.contentString ?? "",
                completion: self.emptyCompletion)
            self.attachCueListeners(track: track, cue: event.cue)
        })
        textTrackRemoveCueListener = track.addEventListener(type: TextTrackEventTypes.REMOVE_CUE, listener: { event in
            self.flutterTextTracksAPI.onTextTrackRemoveCue(textTrackUid: Int64(track.uid), cueUid: Int64(event.cue.uid), completion: self.emptyCompletion)
        })
        textTrackEnterCueListener = track.addEventListener(type: TextTrackEventTypes.ENTER_CUE, listener: { event in
            self.flutterTextTracksAPI.onTextTrackEnterCue(textTrackUid: Int64(track.uid), cueUid: Int64(event.cue.uid), completion: self.emptyCompletion)
        })
        textTrackExitCueListener = track.addEventListener(type: TextTrackEventTypes.EXIT_CUE, listener: { event in
            self.flutterTextTracksAPI.onTextTrackExitCue(textTrackUid: Int64(track.uid), cueUid: Int64(event.cue.uid), completion: self.emptyCompletion)
        })
        textTrackCueChangeListener = track.addEventListener(type: TextTrackEventTypes.CUE_CHANGE, listener: { event in
            self.flutterTextTracksAPI.onTextTrackCueChange(textTrackUid: Int64(track.uid), completion: self.emptyCompletion)
        })
    }
    
    private func removeTrackListeners(track: TextTrack) {
        track.removeEventListener(type: TextTrackEventTypes.ADD_CUE, listener: textTrackAddCueListener!)
        track.removeEventListener(type: TextTrackEventTypes.REMOVE_CUE, listener: textTrackRemoveCueListener!)
        track.removeEventListener(type: TextTrackEventTypes.ENTER_CUE, listener: textTrackEnterCueListener!)
        track.removeEventListener(type: TextTrackEventTypes.EXIT_CUE, listener: textTrackExitCueListener!)
        track.removeEventListener(type: TextTrackEventTypes.CUE_CHANGE, listener: textTrackCueChangeListener!)
    }
    
    private func attachCueListeners(track: TextTrack, cue: TextTrackCue) {
        cueEnterListener = cue.addEventListener(type: TextTrackCueEventTypes.ENTER, listener: {event in
            self.flutterTextTracksAPI.onCueEnter(textTrackUid: Int64(track.uid), cueUid: Int64(cue.uid), completion: self.emptyCompletion)
        })
        cueExitListener = cue.addEventListener(type: TextTrackCueEventTypes.EXIT, listener: {event in
            self.flutterTextTracksAPI.onCueExit(textTrackUid: Int64(track.uid), cueUid: Int64(cue.uid), completion: self.emptyCompletion)
        })
        cueUpdateListener = cue.addEventListener(type: TextTrackCueEventTypes.UPDATE, listener: {event in
            self.flutterTextTracksAPI.onCueUpdate(
                textTrackUid: Int64(track.uid),
                cueUid: Int64(cue.uid),
                endTime: event.cue.endTime ?? 0,
                content: event.cue.contentString ?? "",
                completion: self.emptyCompletion)
        })
    }
    
    private func removeCueListeners(track: TextTrack, cue: TextTrackCue) {
        cue.removeEventListener(type: TextTrackCueEventTypes.ENTER, listener: cueEnterListener!)
        cue.removeEventListener(type: TextTrackCueEventTypes.EXIT, listener: cueExitListener!)
        cue.removeEventListener(type: TextTrackCueEventTypes.UPDATE, listener: cueUpdateListener!)
    }
    
    func setMode(textTrackUid: Int64, mode: TextTrackMode) {
        for i in 0..<theoplayer.textTracks.count {
            var track = theoplayer.textTracks.get(i)
            if (track.uid == textTrackUid) {
                track.mode = TrackTransformer.toTextTrackMode(mode: mode)
            }
        }
    }
    
}
